kubernetes-exec-helper/src/Kubernetes/WSClient.hs

```diff
@@ -0,0 +1,338 @@
+import Control.Concurrent.STM
+import Control.Exception.Safe
+import Control.Monad (forever)
+import Control.Monad.Reader
+import Data.Maybe
+import Data.Monoid ((<>))
+import Data.Text (Text)
+import Kubernetes.K8SChannel
+import Kubernetes.KubeConfig
+import Network.HTTP.Base (urlEncodeVars)
+import Network.Socket(withSocketsDo)
+import qualified Data.Text as T
+import qualified Data.Text.IO as T
+import qualified Network.WebSockets as WS
+import System.Timeout (timeout) 
+import System.IO (hSetBuffering, BufferMode(..), stdin)
+
+{- | 
+  ClientState" maintains all threads that are running,  
+  a writer channel to send messages to the server 
+  and a list of all "(ChannelId, TChan Text)" pairs.
+  Clients can wait on '[Async ThreadId]' and proceed to work with each channel.
+-}
+newtype ClientState a = ClientState {
+    _unState :: 
+      ([Async ThreadId], TChan a, [(ChannelId, TChan a)])
+    }
+
+-- A flag from the python library.
+type PreloadContent = Bool
+
+-- | Secure web sockets connection?
+data Protocol = 
+  WS -- ^ "ws://abc.co" 
+  | WSS -- ^ "wss://abc.co"
+
+instance Show Protocol where 
+  show WS = "ws" 
+  show WSS = "wss"
+
+-- | The host.
+type Host = String
+-- | The port. 
+type Port = Int 
+
+-- | The URL with "Protocol", "Host" and "Port"
+newtype URL = URL {_unP :: (Protocol, Host, Port)} 
+
+-- | The kube config.
+type KubeConfig = Config
+
+-- | Command contains the "Executable" and a list of "Arguments"
+type Command = String
+
+-- | A reader configuration when running the client.
+data ExecClientConfig = 
+  ExecClientConfig {
+  _kubeConfig :: KubeConfig
+  , _url :: URL 
+  , _timeout :: Maybe TimeoutInterval
+  , _preload :: Bool
+  , _commands :: Command
+  } 
+  
+newtype KubernetesClientApp a = 
+  KubernetesClientApp 
+    {runA :: ReaderT ExecClientConfig IO a }
+    deriving (Monad, MonadIO, Functor, Applicative
+      , MonadReader ExecClientConfig)
+
+-- | The core application when a user attaches a command to the pod.
+runApp :: KubeConfig -> Protocol -> Host -> Port -> 
+            Maybe TimeoutInterval -> 
+            PreloadContent -> Command -> IO () 
+runApp kC proto host port timeoutInterval preloadContent command = do
+  let config = ExecClientConfig kC
+                      (URL (proto, host, port)) 
+                      timeoutInterval
+                      preloadContent
+                      command
+  runReaderT (runA exec) config
+
+-- | Read the text from the channels and direct to the appropriate 
+-- | local channel.
+writeToLocalChannels :: [(ChannelId, TChan Text)] -> IO [Async ThreadId]
+writeToLocalChannels channels = 
+  mapM (flip writeToLocalChannel channels) $ [StdIn, StdOut, StdErr] 
+
+-- | Write to a local channel. 
+-- | === Note : The word local channel is used to represent the 
+-- | command line from which a user is running a command.
+writeToLocalChannel :: ChannelId -> [(ChannelId, TChan Text)] -> IO (Async ThreadId)
```

Comment:
Change the order of the first two parameters so you don't need to `flip` in `writeToLocalChannels`?